staging/src/k8s.io/api/policy/v1/types.go

```diff
@@ -0,0 +1,110 @@
+
+	// DisruptedPods contains information about pods whose eviction was
+	// processed by the API server eviction subresource handler but has not
+	// yet been observed by the PodDisruptionBudget controller.
+	// A pod will be in this map from the time when the API server processed the
+	// eviction request to the time when the pod is seen by PDB controller
+	// as having been marked for deletion (or after a timeout). The key in the map is the name of the pod
+	// and the value is the time when the API server processed the eviction request. If
+	// the deletion didn't occur and a pod is still there it will be removed from
+	// the list automatically by PodDisruptionBudget controller after some time.
+	// If everything goes smooth this map should be empty for the most of the time.
+	// Large number of entries in the map may indicate problems with pod deletions.
+	// +optional
+	DisruptedPods map[string]metav1.Time `json:"disruptedPods,omitempty" protobuf:"bytes,2,rep,name=disruptedPods"`
+
+	// Number of pod disruptions that are currently allowed.
+	DisruptionsAllowed int32 `json:"disruptionsAllowed" protobuf:"varint,3,opt,name=disruptionsAllowed"`
+
+	// current number of healthy pods
+	CurrentHealthy int32 `json:"currentHealthy" protobuf:"varint,4,opt,name=currentHealthy"`
+
+	// minimum desired number of healthy pods
+	DesiredHealthy int32 `json:"desiredHealthy" protobuf:"varint,5,opt,name=desiredHealthy"`
+
+	// total number of pods counted by this disruption budget
+	ExpectedPods int32 `json:"expectedPods" protobuf:"varint,6,opt,name=expectedPods"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.5
+// +k8s:prerelease-lifecycle-gen:deprecated=1.22
+
+// PodDisruptionBudget is an object to define the max disruption that can be caused to a collection of pods
+type PodDisruptionBudget struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired behavior of the PodDisruptionBudget.
+	// +optional
+	Spec PodDisruptionBudgetSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+	// Most recently observed status of the PodDisruptionBudget.
+	// +optional
+	Status PodDisruptionBudgetStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.5
+// +k8s:prerelease-lifecycle-gen:deprecated=1.22
+
+// PodDisruptionBudgetList is a collection of PodDisruptionBudgets.
+type PodDisruptionBudgetList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Items           []PodDisruptionBudget `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
```

Comment:
So can we update the version of Eviction since the v1beta1 is part of the pod API (which is GA)? We had a discussion around this in my previous PR to take PDB to GA: https://github.com/kubernetes/kubernetes/pull/81571#discussion_r318874136 and https://github.com/kubernetes/kubernetes/pull/81571#discussion_r318874373

Reply:
We can't change the returned value of the subresource, and most subresources return a specific API type, which is why I thought we couldn't change it. 

The eviction subresource returns v1 Status however, so I *think* we can additionally accept a v1 Eviction type as long as we continue accepting v1beta1 Evictions as well (both v1 and v1beta1 would have to define conversions to the internal Eviction type).

We'd definitely need a test ensuring that both remained accepted.

Reply:
That makes sense. I've tried to look into this, and I'm struggling to see how we would handle this in the `PodExpansion` interface: https://github.com/kubernetes/kubernetes/blob/27c89b9aec73f66529a497910c460af2b25ab6dd/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/pod_expansion.go#L33
I assume we can't add a separate function on the interface to handle the v1 version of the `Eviction`, so which type would we use here? Or should we be able to just keep the `v1beta1` type here and the `v1` version would be converted?

Reply:
> I assume we can't add a separate function on the interface to handle the v1 version of the `Eviction`

That's exactly what I'd do. I'd probably add EvictV1 and EvictV1beta1, document EvictV1 to only apply to 1.21+, and mark `Evict` as deprecated to poke people to explicitly select a version (but not remove it any time soon).

Reply:
Updated the PR so we can accept both v1 Eviction types and v1beta1 Eviction types. This is done by adding additional `pods/evictionv1` and `pods/evictionv1beta1` endpoints and updating the `PodExpansion` with corresponding functions. It also adds test to verify that all endpoints works.

Reply:
Hmm... I was hoping `pods/eviction` could accept either `policy/v1` or `policy/v1beta1` Evictions. I can believe that would not play nicely with the apiserver decoding stack, though. And currently, we can only advertise once accepted version in discovery.

@deads2k @lavalamp, any thoughts on what we should do here? To my knowledge, this is the only subresource of a GA endpoint that accepts a beta version. That's unfortunate to advertise forever and have no graduation path (and I worry a little about how the beta subresource type will interact with the "stop serving deprecated beta types automatically" [changes](https://github.com/kubernetes/kubernetes/pull/99362)).

Reply:
We never had anyone need to change the version of a subresource yet. The plan of record from way back in the day was to permit clients to use the Accept: header to request the version they want to see the result in. We should probably be emiting a warning to clients calling this without specifying an accept header, because such clients would start getting something different if we flip to having v1 as the default.

I think we must continue to accept v1beta1 for quite some time.

Reply:
> I think we must continue to accept v1beta1 for quite some time.

No disagreement there... this is actually a pretty unique resource that is write-only (doesn't support GET and the result of the write is a v1.Status object), so we could conceivably accept both v1beta1 and v1 Eviction objects.

The question is whether we think that subresource accepting both versions is a good thing to aim for, especially if it is likely to involve fixing up bits in the discovery and endpoint handling stack that have single-version fixed assumptions.

Reply:
Oh this is write-only, that's great!

Sadly I don't see any option other than supporting both for some amount of time? I don't think it's OK to abruptly break old clients.

Reply:
So do we want to provide separate endpoints for the different version of the Eviction resource as we currently have in this PR or do we want to see if we can have one endpoint that can accept both the v1 and v1beta1 versions?

If we want to use one endpoint, it seems like that will require changes that are substantial enough that we can't make Eviction v1 in 1.21. So do we want to back out the changes to the Eviction resource from this PR and take just PDBs to GA in 1.21, or do we push everything to 1.22. 1.22 is the automatic deprecation deadline for PDBs?

Reply:
> So do we want to provide separate endpoints for the different version of the Eviction resource as we currently have in this PR or do we want to see if we can have one endpoint that can accept both the v1 and v1beta1 versions?

One endpoint that accepts both would be my preference.

> If we want to use one endpoint, it seems like that will require changes that are substantial enough that we can't make Eviction v1 in 1.21. So do we want to back out the changes to the Eviction resource from this PR and take just PDBs to GA in 1.21, or do we push everything to 1.22.

I'm fine with moving out work on promoting eviction to v1 to 1.22.

> 1.22 is the automatic deprecation deadline for PDBs?

That's when their deprecation period starts. It ends in 1.25 and we stop serving them then.

Reply:
> I'm fine with moving out work on promoting eviction to v1 to 1.22.

I was told we can't GA from a deprecated version, and we'd need to go through a 2nd beta, isn't that true?

Reply:
If 1.22 adds Eviction v1 and marks Eviction v1beta1 deprecated, that seems reasonable. Eviction v1beta1 is an oddball since it's a subresource of a v1 API, so I expect we'll have interesting things to iron out there anyway.

Reply:
I have backed out the changes for updating Eviction to v1.

Reply:
thanks, can you make sure this discussion is captured in an issue associated with the KEP, so we can make sure this happens in 1.22?

Reply:
Added a comment on the tracking issue: https://github.com/kubernetes/enhancements/issues/85#issuecomment-791591925